dd-sdk-android/src/main/kotlin/com/datadog/android/rum/webview/RumEventConsumer.kt

```diff
@@ -0,0 +1,137 @@
+import com.datadog.android.rum.model.ResourceEvent
+import com.datadog.android.rum.model.ViewEvent
+import com.google.gson.JsonObject
+import com.google.gson.JsonParseException
+
+internal class RumEventConsumer(
+    private val dataWriter: DataWriter<Any>,
+    private val timeProvider: TimeProvider
+) {
+
+    private var currentOffset = timeProvider.getServerOffsetMillis()
+
+    fun consume(event: JsonObject, eventType: String) {
+        if (!isSessionDropped()) {
+            val mappedEvent = map(event, eventType)
+            if (mappedEvent != null) {
+                dataWriter.write(mappedEvent)
+            }
+        }
+    }
+
+    private fun map(event: JsonObject, eventType: String): JsonObject? {
+        val rumContext = GlobalRum.getRumContext()
+
+        return try {
+            when (eventType) {
+                VIEW_EVENT_TYPE -> {
+                    currentOffset = timeProvider.getServerOffsetMillis()
+                    ViewEvent.fromJson(event.toString()).mapToMobile(rumContext, currentOffset)
+                }
+                ACTION_EVENT_TYPE -> {
+                    ActionEvent.fromJson(event.toString()).mapToMobile(rumContext, currentOffset)
+                }
+                RESOURCE_EVENT_TYPE -> {
+                    ResourceEvent.fromJson(event.toString()).mapToMobile(rumContext, currentOffset)
+                }
+                ERROR_EVENT_TYPE -> {
+                    ErrorEvent.fromJson(event.toString()).mapToMobile(rumContext, currentOffset)
+                }
+                LONG_TASK_EVENT_TYPE -> {
+                    LongTaskEvent.fromJson(event.toString()).mapToMobile(rumContext, currentOffset)
+                }
+                else -> {
+                    null
+                }
+            }
+        } catch (e: JsonParseException) {
+            sdkLogger.e(JSON_PARSING_ERROR_MESSAGE, e)
+            null
+        }
+    }
+
+    private fun isSessionDropped(): Boolean {
+        val appScope = (GlobalRum.get() as? DatadogRumMonitor)?.rootScope as? RumApplicationScope
+            ?: return false
+        return (appScope.childScope as? RumSessionScope)?.keepSession ?: false
+    }
+
+    private fun ViewEvent.mapToMobile(context: RumContext, timeOffset: Long): JsonObject {
+        val copy = this.copy(
+            application = this.application.copy(id = context.applicationId),
+            session = this.session.copy(id = context.sessionId),
+            service = CoreFeature.serviceName,
+            date = this.date + timeOffset
+        )
+        return copy.toJson().asJsonObject
+    }
+
+    private fun ActionEvent.mapToMobile(context: RumContext, timeOffset: Long): JsonObject {
+        val copy = this.copy(
+            application = this.application.copy(id = context.applicationId),
+            session = this.session.copy(id = context.sessionId),
+            service = CoreFeature.serviceName,
+            date = this.date + timeOffset
```

Comment:
`+` we still need to solve the problem with `timeOffset` being impacted by NTP time sync, right?

Reply:
if the events order is being kept we can go with this simpler way by just reseting the `timeOffset` when a new `ViewEvent` comes in. If we can have multiple events in the same time then indeed we need to rethink this and go with the `Map<ViewId,Offset>` approach